rfc/ipni-w3c.md

```diff
@@ -0,0 +1,346 @@
+    E --> F(ipni Advertisement content ƛ)
+    F --> |PUT /advertCid|I
+    F --> |advert CID| G[/Advertisement queue/]
+    G --> H(ipni publish ƛ)
+    H --> |PUT /head|I[(Advert Bucket)]
+    H --> |POST head|IPNI[["`**IPNI**`"]]
+    
+    carpark[(carpark\nbucket)] --> |ObjectCreated fa:fa-car|w3infra-carpark-consumer(carpark-consumer ƛ)
+    w3infra-carpark-consumer -->|region/bucket/cid/cid.car| C[/indexer queue/]
+    indexer ---> dynamo[Dynamo\nblocks index]
+```
+
+## Proposal
+Provide a ipni/offer UCAN ability to submit the location of a CAR and a CARx-index that a user has stored with w3s, to make the them discoverable via IPFS implementations and other IPNI consumers.
+
+The ipni/offer is optional, and if not invoked the user's data is not indexed.
+
+### Indexing stored data
+
+```mermaid
+sequenceDiagram
+    actor Alice
+    Alice->>w3s: put CAR + CARx-index
+    activate Alice
+    Alice-->>Alice: create location claims
+    Alice->>w3s: ipni/offer (inc, loc claims)
+    deactivate Alice
+    activate w3s
+    w3s-->>w3s: fetch & verify CARx-index
+    w3s-->>w3s: create/store claims bucket
+    w3s-->>w3s: write advertisement
+    w3s-)ipni: publish head (CID)
+    deactivate w3s
+    ipni-->>w3s: fetch advert
+    activate ipni
+    ipni-->>ipni: index entries
+    deactivate ipni
+```
+Invoke `ipni/offer` with the CID for an [inclusion claim] that associates a CAR CID with a CARx-index CID.
+
+**UCAN invocation** example
+
+```json
+{
+  "iss": "did:key:zAlice",
+  "aud": "did:web:web3.storage",
+  "att": [{
+    "can": "ipni/offer",
+    "with": "did:key:space", // users space DID
+    "nb": {
+        "inclusion": CID   // inclusion claim CID
+    }
+  }]
+}
+```
+
+**Location claims** example
+
+```json
+{
+  "op": "assert/location",
+  "rsc": "https://web3.storage",
+  "input": {
+    "content" : CID /* CAR CID */, 
+    "location": "https://r2.cf/bag...car"
+  }
+}
+```
+```json
+{
+  "op": "assert/location",
+  "rsc": "https://web3.storage",
+  "input": {
+    "content" : CID /* CARx-index CID */, 
+    "location": "https://r2.cf/bag...carx"
+  }
+}
+```
+
+**Inclusion claim** example
+
+```json
+{
+  "content": CID, // CAR CID
+  "includes": CID // CARx-index Index CID
+}
+```
+
+When `ipni/offer` is invoked, encoded inclusion claim and encoded location claims for the CAR and the CARx-index are all sent with the invocation.
+
+_Alternatively, the w3s service could create the location claims. The user would still need to provide the locations, so might as well do this as a location claim. If w3s moves the data, then it can replace the location claims with new ones_
+
+The service must fetch he CARx-index and parse it to find the set of multihashes included in the CAR. see: [Verifying the CARv2 Index](#verifying-the-carv2-index)
+
+_Alternatively the service could generate the CARx-index and store it, but it would be better to have the clients do that work of generating and storing the index._
```

Comment:
Strong push for clients to do this ONLY.

My basic rubric is:
authority over what the data is is ALWAYS the client
authority over where the data lives is ALWAYS w3s

Reply:
I should add I think we should absolutely make the entire ipni/offer process, plus generating the CAR-index, etc, entirely optional (front client software perspective, it's on by default to start). This leaves the ability to not index things potentially in the clients hands.

This design for example, enables Bitswap via using IPNI to map to CAR CID for any CID. If people don't want Bitswap, potentially eventually with flat data not CAR-ified, they might not need all this. They might just need to index the root and other CIDs of interest.

Reply:
Additionally, this all assumes public retrieval is an intended outcome -- for private retrieval, not advertising at all seems like a given.

Reply:
> assumes public retrieval 

The data could also be encrypted by the client, and the symmetric decryption key could be encrypted for each allowed reader, as is done with [age-encryption](https://github.com/C2SP/C2SP/blob/main/age.md). While that should probably be left up to the client, it might be nice for the client to have a feature like that built-in.